### PR TITLE
Spdp::fini_bit waits

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

Waiting on a ReactorInterceptor::Command can cause deadlock (see

Solution
--------

Factor out a new class called BitSubscriber with atomic methods.

### CODE REVIEW

File: dds/DCPS/BuiltInTopicUtils.h [hunk 1]
```diff
@@ -222,6 +222,67 @@ DDS::BuiltinTopicKey_t keyFromSample(TopicType*)
   return key;
 }
 
+class OpenDDS_Dcps_Export BitSubscriber : public RcObject {
+public:
+  BitSubscriber()
+  {}
+
+  BitSubscriber(const DDS::Subscriber_var& bit_subscriber)
```
Comment (mitza-oci) [inline dds/DCPS/BuiltInTopicUtils.h:230]:
```suggestion
  explicit BitSubscriber(const DDS::Subscriber_var& bit_subscriber)
```

File: dds/DCPS/BuiltInTopicUtils.h [hunk 2]
```diff
@@ -222,6 +222,67 @@ DDS::BuiltinTopicKey_t keyFromSample(TopicType*)
...
+  DDS::InstanceHandle_t add_connection_record(const ConnectionRecord& cr,
+                                              DDS::ViewStateKind view_state);
+  void remove_connection_record(const ConnectionRecord& cr);
+
+  DDS::InstanceHandle_t add_thread_status(const InternalThreadBuiltinTopicData& ts,
+                                          DDS::ViewStateKind view_state,
+                                          const SystemTimePoint& timestamp);
+  void remove_thread_status(const InternalThreadBuiltinTopicData& ts);
+
+  void bit_pub_listener_hack(DomainParticipantImpl* participant);
```
Comment (mitza-oci) [inline dds/DCPS/BuiltInTopicUtils.h:277]: add comments addressing why it's a hack

File: dds/DCPS/BuiltInTopicUtils.cpp [hunk 3]
```diff
@@ -34,6 +38,418 @@ const char* const BUILT_IN_CONNECTION_RECORD_TOPIC_TYPE = "CONNECTION_RECORD_BUI
...
+  if (!bit_subscriber_) {
+    return;
+  }
+
+  DDS::DataReader_var dr =
+    bit_subscriber_->lookup_datareader(BUILT_IN_PUBLICATION_TOPIC);
+  DDS::PublicationBuiltinTopicDataDataReader_var bit_pub_dr =
+    DDS::PublicationBuiltinTopicDataDataReader::_narrow(dr);
+
+  if (!CORBA::is_nil(bit_pub_dr.in())) {
```
Comment (mitza-oci) [inline dds/DCPS/BuiltInTopicUtils.cpp:436]:
```suggestion
  if (bit_pub_dr) {
```

File: dds/DCPS/BuiltInTopicUtils.cpp [hunk 4]
```diff
@@ -34,6 +38,418 @@ const char* const BUILT_IN_CONNECTION_RECORD_TOPIC_TYPE = "CONNECTION_RECORD_BUI
...
+    DDS::DataReaderListener_var listener = bit_pub_dr->get_listener();
+    if (CORBA::is_nil(listener.in())) {
```
Comment (mitza-oci) [inline dds/DCPS/BuiltInTopicUtils.cpp:438]:
```suggestion
    if (listener) {
```

File: dds/DCPS/BuiltInTopicUtils.cpp [hunk 5]
```diff
@@ -34,6 +38,418 @@ const char* const BUILT_IN_CONNECTION_RECORD_TOPIC_TYPE = "CONNECTION_RECORD_BUI
...
+      DDS::DataReaderListener_var bit_pub_listener =
+        new BitPubListenerImpl(participant);
+      bit_pub_dr->set_listener(bit_pub_listener, DDS::DATA_AVAILABLE_STATUS);
+      // Must call on_data_available when attaching a listener late - samples may be waiting
+      SubscriberImpl* sub = dynamic_cast<SubscriberImpl*>(bit_subscriber_.in());
+      DataReaderImpl* reader = dynamic_cast<DataReaderImpl*>(bit_pub_dr.in());
+      TheServiceParticipant->job_queue()->enqueue(make_rch<DataReaderImpl::OnDataAvailable>(rchandle_from(sub), bit_pub_listener, rchandle_from(reader), true, false, false));
```
Comment (mitza-oci) [inline dds/DCPS/BuiltInTopicUtils.cpp:445]: Does the `OnDataAvailable` constructor need both the reader and the subscriber?  Each reader already knows which subscriber it belongs to.

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/LinuxNetworkConfigMonitor.h [hunk 6]
```diff
@@ -37,33 +37,43 @@ class OpenDDS_Dcps_Export LinuxNetworkConfigMonitor
...
       : lncm_(lncm)
     {}
 
   private:
     void execute()
     {
-      if (reactor()->register_handler(lncm_, READ_MASK) != 0) {
+      RcHandle<LinuxNetworkConfigMonitor> lncm = lncm_.lock();
+      if (!lncm) {
+	return;
```
Comment (iguessthislldo) [inline dds/DCPS/LinuxNetworkConfigMonitor.h:49]:
```suggestion
	    return;
```

File: dds/DCPS/LinuxNetworkConfigMonitor.h [hunk 7]
```diff
@@ -37,33 +37,43 @@ class OpenDDS_Dcps_Export LinuxNetworkConfigMonitor
...
+    RemoveHandler(WeakRcHandle<LinuxNetworkConfigMonitor> lncm)
       : lncm_(lncm)
     {}
 
     void execute()
     {
-      reactor()->remove_handler(lncm_, READ_MASK);
+      RcHandle<LinuxNetworkConfigMonitor> lncm = lncm_.lock();
+      if (!lncm) {
+	return;
```
Comment (iguessthislldo) [inline dds/DCPS/LinuxNetworkConfigMonitor.h:70]:
```suggestion
	    return;
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/RTPS/Sedp.h [hunk 8]
```diff
@@ -367,105 +367,105 @@ struct DiscoveredParticipant {
 #endif
 };
 
-class Sedp : public DCPS::RcEventHandler {
-private:
-  struct DiscoveredSubscription : PoolAllocationBase {
-    DiscoveredSubscription()
+struct DiscoveredPublication : DCPS::PoolAllocationBase {
```
[the remark this answers is not in the document]
Comment (jrw972) [reply dds/DCPS/RTPS/Sedp.h:370]: I'm wondering why this changed in the first place.

**Review by mitza-oci — APPROVED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**